xdsl/backend/block_throughput_cost_model.py

```diff
@@ -0,0 +1,127 @@
+"""
+This module defines an abstract base class `BlockThroughputCostModel` for estimating the
+throughput of a given `Block`.
+
+It includes a concrete implementation, `MCABlockThroughputCostModel`, that uses
+the `llvm-mca` tool to perform this estimation.
+"""
+
+import os
+import subprocess
+from abc import ABC, abstractmethod
+from shutil import which
+from tempfile import NamedTemporaryFile
+
+from xdsl.backend.assembly_printer import AssemblyPrintable, AssemblyPrinter
+from xdsl.ir import Block
+
+
+class BlockThroughputCostModel(ABC):
+    """
+    An abstract base class for a throughput cost model.
+
+    This class provides an interface for estimating the throughput of a block
+    of assembly-like operations for a specific microarchitecture.
+    """
+
+    arch: str
+    src_path: str
+
+    def __init__(self, arch: str):
+        self.arch = arch
+
+    @abstractmethod
+    def name(self) -> str:
+        pass
```

Comment:
Let's move these to the `ExternalBlockThroughputCostModel`? Also, it would be nice to make it clear that `name` is actually the name of the command-line tool to call, maybe `bin_name`? Or `tool_name`? Not sure what's best but something more explicit would be nice.